File: code/dart/rest/server/lib/services/task_service.dart
import '../models/task.dart';

class TaskService {
  final Map<String, Task> _tasks = {};

  TaskService() {
    _initializeSampleData();
  }

  void _initializeSampleData() {
    final sampleTasks = [
      Task.create(
        title: 'Implement Dart REST API',
        description: 'Create REST API with Shelf framework',
        priority: TaskPriority.high,
        tags: ['dart', 'rest', 'api'],
        assignedTo: 'developer',
      ),
      Task.create(
        title: 'Add Flutter client',
        description: 'Create Flutter app that consumes the API',
        priority: TaskPriority.medium,
        tags: ['dart', 'flutter', 'mobile'],
        assignedTo: 'developer',
      ),
    ];

    for (final task in sampleTasks) {
      _tasks[task.id] = task;
    }
  }

  Future<ListTasksResponse> listTasks({
    int? pageSize,
    String? pageToken,
    TaskStatus? status,
    String? assignedTo,
    List<String>? tags,
    SortOrder? sortOrder,
  }) async {
    var tasks = _tasks.values.toList();

    // Apply filters
    if (status != null) {
      tasks = tasks.where((task) => task.status == status).toList();
    }

    if (assignedTo != null) {
      tasks = tasks.where((task) => task.assignedTo == assignedTo).toList();
    }

    if (tags != null && tags.isNotEmpty) {
      tasks = tasks.where((task) {
        return tags.every((tag) => task.tags.contains(tag));
      }).toList();
    }

    // Apply sorting
    tasks = _sortTasks(tasks, sortOrder);

    // Apply pagination
    final effectivePageSize = (pageSize ?? 20).clamp(1, 100);
    final startIndex = pageToken != null ? int.tryParse(pageToken) ?? 0 : 0;
    final endIndex = startIndex + effectivePageSize;

    final paginatedTasks = tasks.skip(startIndex).take(effectivePageSize).toList();
    final nextPageToken = endIndex < tasks.length ? endIndex.toString() : null;

    return ListTasksResponse(
      tasks: paginatedTasks,
      nextPageToken: nextPageToken,
      totalCount: tasks.length,
    );
  }

  Future<Task> getTask(String id) async {
    final task = _tasks[id];
    if (task == null) {
      throw NotFoundException('Task with ID $id not found');
    }
    return task;
  }

  Future<Task> createTask(CreateTaskRequest request) async {
    final task = Task.create(
      title: request.title,
      description: request.description ?? '',
      priority: request.priority ?? TaskPriority.medium,
      tags: request.tags ?? [],
      assignedTo: request.assignedTo,
      dueDate: request.dueDate,
    );

    _tasks[task.id] = task;
    return task;
  }

  Future<Task> updateTask(String id, UpdateTaskRequest request) async {
    final task = await getTask(id);

    DateTime? completedAt = task.completedAt;
    if (request.status == TaskStatus.completed && 
        task.status != TaskStatus.completed) {
      completedAt = DateTime.now();
    }

    final updatedTask = task.copyWith(
      title: request.title,
      description: request.description,
      status: request.status,
      priority: request.priority,
      tags: request.tags,
      assignedTo: request.assignedTo,
      dueDate: request.dueDate,
      completedAt: completedAt,
    );

    _tasks[id] = updatedTask;
    return updatedTask;
  }

  Future<void> deleteTask(String id) async {
    final task = await getTask(id);
    _tasks.remove(id);
  }

  Future<Task> updateTaskStatus(String id, TaskStatus status) async {
    return updateTask(id, UpdateTaskRequest(status: status));
  }

  List<Task> _sortTasks(List<Task> tasks, SortOrder? sortOrder) {
    if (sortOrder == null) return tasks;

    switch (sortOrder) {
      case SortOrder.createdAtAsc:
        tasks.sort((a, b) => a.createdAt.compareTo(b.createdAt));
        break;
      case SortOrder.createdAtDesc:
        tasks.sort((a, b) => b.createdAt.compareTo(a.createdAt));
        break;
      case SortOrder.dueDateAsc:
        tasks.sort((a, b) {
          if (a.dueDate == null) return 1;
          if (b.dueDate == null) return -1;
          return a.dueDate!.compareTo(b.dueDate!);
        });
        break;
      case SortOrder.dueDateDesc:
        tasks.sort((a, b) {
          if (a.dueDate == null) return 1;
          if (b.dueDate == null) return -1;
          return b.dueDate!.compareTo(a.dueDate!);
        });
        break;
      case SortOrder.priorityAsc:
        tasks.sort((a, b) => _getPriorityValue(a.priority)
            .compareTo(_getPriorityValue(b.priority)));
        break;
      case SortOrder.priorityDesc:
        tasks.sort((a, b) => _getPriorityValue(b.priority)
            .compareTo(_getPriorityValue(a.priority)));
        break;
    }

    return tasks;
  }

  int _getPriorityValue(TaskPriority priority) {
    switch (priority) {
      case TaskPriority.low:
        return 1;
      case TaskPriority.medium:
        return 2;
      case TaskPriority.high:
        return 3;
      case TaskPriority.critical:
        return 4;
    }
  }
}

class NotFoundException implements Exception {
  final String message;
  NotFoundException(this.message);

  @override
  String toString() => message;
}

class ValidationException implements Exception {
  final String message;
  ValidationException(this.message);

  @override
  String toString() => message;
}

File: code/javascript/rest/server/services/taskService.js
import { v4 as uuidv4 } from 'uuid';

export class TaskService {
  constructor() {
    this.tasks = new Map();
    this.initializeSampleData();
  }

  initializeSampleData() {
    const sampleTasks = [
      {
        title: 'Complete JavaScript chapter',
        description: 'Write comprehensive REST and gRPC examples',
        priority: 'high',
        status: 'in_progress',
        tags: ['book', 'javascript', 'api'],
        assigned_to: 'david'
      },
      {
        title: 'Review TypeScript patterns',
        description: 'Ensure TypeScript examples follow best practices',
        priority: 'medium',
        status: 'pending',
        tags: ['book', 'typescript', 'review'],
        assigned_to: 'david'
      }
    ];

    sampleTasks.forEach(task => this.createTask(task));
  }

  async listTasks(filters = {}) {
    let tasks = Array.from(this.tasks.values());
    
    // Apply filters
    if (filters.status) {
      tasks = tasks.filter(task => task.status === filters.status);
    }
    
    if (filters.assigned_to) {
      tasks = tasks.filter(task => task.assigned_to === filters.assigned_to);
    }
    
    if (filters.tags && filters.tags.length > 0) {
      const filterTags = filters.tags.split(',').map(t => t.trim());
      tasks = tasks.filter(task => 
        filterTags.every(tag => task.tags.includes(tag))
      );
    }
    
    // Apply sorting
    tasks = this.sortTasks(tasks, filters.sort_order);
    
    // Apply pagination
    const page_size = Math.min(filters.page_size || 20, 100);
    const start_index = filters.page_token ? parseInt(filters.page_token) : 0;
    const paginatedTasks = tasks.slice(start_index, start_index + page_size);
    
    return {
      tasks: paginatedTasks,
      next_page_token: start_index + page_size < tasks.length 
        ? String(start_index + page_size) 
        : null,
      total_count: tasks.length
    };
  }

  async getTask(id) {
    const task = this.tasks.get(id);
    if (!task) {
      throw new NotFoundError(`Task with ID ${id} not found`);
    }
    return task;
  }

  async createTask(taskData) {
    const now = new Date().toISOString();
    const task = {
      id: uuidv4(),
      title: taskData.title,
      description: taskData.description || '',
      status: taskData.status || 'pending',
      priority: taskData.priority || 'medium',
      tags: taskData.tags || [],
      created_by: taskData.created_by || 'system',
      assigned_to: taskData.assigned_to || null,
      created_at: now,
      updated_at: now,
      due_date: taskData.due_date || null,
      completed_at: null
    };
    
    this.tasks.set(task.id, task);
    return task;
  }

  async updateTask(id, updates) {
    const task = await this.getTask(id);
    
    const updatedTask = {
      ...task,
      ...updates,
      id: task.id, // Prevent ID changes
      created_at: task.created_at, // Preserve creation time
      updated_at: new Date().toISOString()
    };
    
    // Handle status change to completed
    if (updates.status === 'completed' && task.status !== 'completed') {
      updatedTask.completed_at = new Date().toISOString();
    }
    
    this.tasks.set(id, updatedTask);
    return updatedTask;
  }

  async deleteTask(id) {
    const task = await this.getTask(id);
    this.tasks.delete(id);
    return task;
  }

  sortTasks(tasks, sortOrder) {
    switch (sortOrder) {
      case 'created_at_asc':
        return tasks.sort((a, b) => new Date(a.created_at) - new Date(b.created_at));
      case 'created_at_desc':
        return tasks.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
      case 'due_date_asc':
        return tasks.sort((a, b) => {
          if (!a.due_date) return 1;
          if (!b.due_date) return -1;
          return new Date(a.due_date) - new Date(b.due_date);
        });
      case 'due_date_desc':
        return tasks.sort((a, b) => {
          if (!a.due_date) return 1;
          if (!b.due_date) return -1;
          return new Date(b.due_date) - new Date(a.due_date);
        });
      case 'priority_asc':
        return tasks.sort((a, b) => this.priorityValue(a.priority) - this.priorityValue(b.priority));
      case 'priority_desc':
        return tasks.sort((a, b) => this.priorityValue(b.priority) - this.priorityValue(a.priority));
      default:
        return tasks;
    }
  }

  priorityValue(priority) {
    const values = { low: 1, medium: 2, high: 3, critical: 4 };
    return values[priority] || 2;
  }
}

export class NotFoundError extends Error {
  constructor(message) {
    super(message);
    this.name = 'NotFoundError';
    this.statusCode = 404;
  }
}

export class ValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ValidationError';
    this.statusCode = 400;
  }
}